src/api.ts

```diff
@@ -539,39 +568,127 @@ function createApi({
       const offlineStyle: OfflineStyle = {
         ...(await uncompositeStyle(style)),
         id: styleId,
-        sources: await createOfflineSources(style.sources),
+        sources: await createOfflineSources({
+          accessToken,
+          styleId,
+          sources: style.sources,
+        }),
       }
 
       db.prepare(
         'INSERT INTO Style (id, stylejson) VALUES (:id, :stylejson)'
       ).run({ id: styleId, stylejson: JSON.stringify(offlineStyle) })
 
+      // Create the records in the joins table for TilesetsOnStyles
+      // TODO: How to insert multiple records at once?
+      Object.values(offlineStyle.sources).forEach((offlineSource) => {
+        db.prepare<{ tilesetId: string; styleId: string }>(
+          'INSERT INTO TilesetsOnStyles (tilesetId, styleId) VALUES (:tilesetId, :styleId)'
+        ).run({ tilesetId: offlineSource.tilesetId, styleId })
+      })
+
       return addOfflineUrls(offlineStyle)
     },
 
+    // TODO: May need to accept an access token
     async putStyle(id, style) {
-      db.prepare(
-        'INSERT INTO Style (id, stylejson) VALUES (:id, :stylejson)'
-      ).run({ id, stylejson: style })
-      return { ...style, id }
+      if ('id' in style && id !== style.id) {
+        throw new MismatchedIdError(id, style.id)
+      }
+
+      if (!styleExists(id)) {
+        throw new NotFoundError(id)
+      }
+
+      const offlineStyle: OfflineStyle = {
+        ...(await uncompositeStyle(style)),
+        id,
+        // TODO: Is this the right thing to do? May need to update createOfflineSources to handle pre-existing tilesets for this style
+        sources: await createOfflineSources({
+          styleId: id,
+          sources: style.sources,
+        }),
+      }
+
+      db.prepare('UPDATE Style SET stylejson = :stylejson WHERE id = :id').run({
+        id,
+        stylejson: JSON.stringify(offlineStyle),
+      })
+
+      // TODO: update TilesetsOnStyles table to reflect any changes in the sources field?
+
+      return addOfflineUrls(offlineStyle)
     },
 
     async listStyles(limit?: number) {
+      const styles: OfflineStyle[] = []
+
       const baseQuery = 'SELECT stylejson FROM Style'
+
       const stmt =
         limit !== undefined
           ? db.prepare(`${baseQuery} LIMIT ?`).bind(limit)
           : db.prepare(baseQuery)
 
-      return stmt
-        .all()
-        .map((row: { stylejson: string }) => JSON.parse(row.stylejson))
+      stmt.all().forEach(({ stylejson }: { stylejson: string }) => {
+        try {
+          const style: OfflineStyle = JSON.parse(stylejson)
+          styles.push(addOfflineUrls(style))
+        } catch (err) {
+          // TODO: What should we do here? e.g. omit or throw?
```

Comment:
I think throw, because this suggests something is broken in the DB. Should report to bugsnag at least.